Add debug_flowHeight API endpoint

janezpodhostnik opened the pull request:
## Description

Added `debug_flowHeight` API endpoint so you can get the flow height at a given EVM block

______

For contributor use:

- [ ] Targeted PR against `master` branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/flow-nft/blob/master/CONTRIBUTING.md#styleguides).
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

janezpodhostnik commented:
@zhangchiqing @m-Peter What do you think of the API endpoint name?

zhangchiqing commented on api/debug.go:
```diff
@@ -361,6 +361,24 @@ func (d *DebugAPI) TraceCall(
 }
 
+// FlowHeight returns the Flow height for the given EVM block.
+func (d *DebugAPI) FlowHeight(
```
```suggestion
func (d *DebugAPI) FlowHeightByEVMBlock(
```

m-Peter replied:
Then we would have to query it by: `eth_flowHeightByEVMBlock`, it's rather easy to get it wrong :sweat_smile:

zhangchiqing replied:
>Then we would have to query it by: eth_flowHeightByEVMBlock, it's rather easy to get it wrong 😅

Why it's easy to get wrong? Sorry, I didn't get it.

zhangchiqing replied:
I see. How do we map a DebugAPI method to a RPC method. Is there a conversion defined somewhere? Or could be specify `FlowHeightByEVMBlock` API will be mapped to `eth_flow_height_by_evm_block` something like that?

m-Peter replied:
Unfortunately, this isn't something that we have control over, it's specific by some reflection that the JSON-RPC server from Geth does.

m-Peter replied:
I was actually wrong, it will be `debug_flowHeightByEVMBlock` and not `eth_*`.

zhangchiqing replied:
OK. Is `TraceBlockByNumber` auto mapped to `debug_traceBlockByNumber`? If so, `debug_flowHeightByEVMBlock` is the same naming convention, right?

m-Peter replied:
Yep, the naming convention is just to prepend the namespace (e.g. `eth_` / `net` / `debug`) and lowercase the first character of the method name. `debug_flowHeightByEVMBlock` is also fine, I believe we will have this command documented, so it's not like we will write it by hand.

m-Peter replied:
Though I am not entirely sure if it will be `debug_flowHeightByEVMBlock` or `debug_flowHeightByEvmBlock`.

zhangchiqing replied:
>Though I am not entirely sure if it will be debug_flowHeightByEVMBlock or debug_flowHeightByEvmBlock.

Oh, I see your concern now. Yeah, let's find out.

I'm ok that we name it to `FlowHeightByBlock` to avoid being mapped to `debug_flowHeightByEvmBlock`, which looks weird. In that case, we better mention EVM block in the comments.

m-Peter replied:
I tried this locally, and it will be mapped to `debug_flowHeightByEVMBlock`, so only the first character of the method name gets lower-cased. The rest remains intact.

zhangchiqing commented on api/debug.go:
```diff
@@ -361,6 +361,24 @@ func (d *DebugAPI) TraceCall(
 	return tracer.GetResult()
 }
 
+// FlowHeight returns the Flow height for the given EVM block.
```
```suggestion
// FlowHeight returns the Flow height for the given EVM block specified either by EVM 
// block height or EVM block hash.
```

m-Peter commented on tests/e2e_web3js_test.go:
```diff
@@ -36,6 +36,10 @@ func TestWeb3_E2E(t *testing.T) {
 		runWeb3Test(t, "debug_traces_test")
 	})
 
+	t.Run("test transaction traces", func(t *testing.T) {
```
```suggestion
	t.Run("test debug utils", func(t *testing.T) {
```

m-Peter commented on tests/web3js/debug_util_test.js:
```diff
@@ -0,0 +1,14 @@
+const helpers = require('./helpers')
+
+it('should retrieve flow height', async () => {
+    response = await helpers.callRPCMethod(
```
```suggestion
    let response = await helpers.callRPCMethod(
```

m-Peter commented on tests/web3js/debug_util_test.js:
```diff
@@ -0,0 +1,14 @@
+    assert.isDefined(response.body)
+
+    let height = response.body.result
+    assert.isTrue(height > 0)
```
```suggestion
    assert.equal(height, conf.startBlockHeight)
```

m-Peter replied:
You can get `conf` with -> `const conf = require('./config')`

m-Peter reviewed (approved):
Awesome :clap:

m-Peter commented:
@janezpodhostnik I forgot to mention, there's also this map here: https://github.com/onflow/flow-evm-gateway/blob/main/api/api.go#L37-L39, so we might want to update it, when we settle on the endpoint's name :pray:

janezpodhostnik commented:
> @janezpodhostnik I forgot to mention, there's also this map here: https://github.com/onflow/flow-evm-gateway/blob/main/api/api.go#L37-L39, so we might want to update it, when we settle on the endpoint's name 🙏

Thanks. Added. 

I also changed the method name as per the current consensus.

m-Peter reviewed (approved):
LGTM!
